sympy/core/relational.py

```diff
@@ -426,16 +427,30 @@ def __new__(cls, lhs, rhs=None, **options):
                     isinstance(rhs, Boolean)):
                 return S.false  # only Booleans can equal Booleans
 
-            # check finiteness
-            fin = L, R = [i.is_finite for i in (lhs, rhs)]
-            if None not in fin:
-                if L != R:
+            if lhs.is_infinite or rhs.is_infinite:
+                if fuzzy_xor([lhs.is_infinite, rhs.is_infinite]):
                     return S.false
-                if L is False:
-                    if lhs == -rhs:  # Eq(oo, -oo)
-                        return S.false
-                    return S.true
-            elif None in fin and False in fin:
+                if fuzzy_xor([lhs.is_extended_real, rhs.is_extended_real]):
+                    return S.false
+                if fuzzy_and([lhs.is_extended_real, rhs.is_extended_real]):
+                    r = fuzzy_xor([lhs.is_extended_positive, fuzzy_not(rhs.is_extended_positive)])
+                    return S(r)
+
+                # Try to split real/imaginary parts and equate them
+                def split_real_imag(expr):
+                    return sift(Add.make_args(expr), lambda a:a.is_extended_real)
+
+                lhs_ri = split_real_imag(lhs)
+                if not lhs_ri[None]:
+                    rhs_ri = split_real_imag(rhs)
+                    if not rhs_ri[None]:
+                        I = S.ImaginaryUnit
+                        eq_real = Eq(Add(*lhs_ri[True]), Add(*rhs_ri[True]))
+                        eq_imag = Eq(I*Add(*lhs_ri[False]), I*Add(*rhs_ri[False]))
+                        res = fuzzy_and(map(fuzzy_bool, [eq_real, eq_imag]))
+                        if res is not None:
+                            return S(res)
```

Comment:
The logic here is valid (in a purely logical sense) but leads to infinite recursion:
```julia
In [1]: x = Symbol('x', infinite=True, extended_real=False)                                                                       

In [2]: y = Symbol('y', infinite=True, extended_real=False)                                                                       

In [3]: Eq(x, y)                                                                                                                  
---------------------------------------------------------------------------
RecursionError
```

Reply:
Actually the logic isn't valid. The code doesn't really split into real and imaginary parts but rather into real and not-purely-real parts so multiplying by `I` doesn't guarantee that we end up with a real equation for the real parts and a real equation for the imaginary parts.

Reply:
Maybe we should check for `(I*t).is_extended_real` when sifting.